Display product details

JDurstberger opened the pull request.

Commit:
Refactor BaseViewModel.compose to use method references

Commit:
Adds empty ProductDetailActivity

Commit:
Adds progressBar to ProductDetailActivity

Commit:
Push disposables into BaseActivity

Commit:
Inject ProductDetailViewModelFactory with dagger2

Commit:
Remove example instrumented test

Commit:
Add screen test for loading behaviour on product detail screen

Commit:
Reformat creation of ProductDetail in stub repository

Commit:
Add screen tests for product detail loading behaviour

Commit:
Render product details on screen

Commit:
Clean up image loading

Commit:
Rename ProductDetailViewState into ProductDetailScreenViewState

Commit:
Rename ProductDetail to ProductDetailViewStaet

Commit:
Update gradle to 5 to support JDK11

Commit:
Move ProductDetailEntity to entities package

Commit:
Rename ProductDetailEntity to ProductDetail

Commit:
Return ProductDetail from success result

Commit:
Fix empty imageUrl for ProductDetail screen test

Commit:
Reorganise ProductDetail classes into multiple files

samcackett commented on app/build.gradle:
```diff
@@ -29,7 +29,8 @@ androidExtensions {
+    implementation project(':data')
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
 
     implementation 'com.android.support:appcompat-v7:27.1.1'
```
I know not all of these are you, but can you add variables for the version numbers instead of writing them out as strings

samcackett commented on app/src/androidTest/java/com/moonpig/mvisample/screentests/di/TestApplicationComponent.kt:
```diff
@@ -0,0 +1,9 @@
+import com.moonpig.mvisample.di.ApplicationComponent
+import dagger.Component
+
+@Component(modules = [
```
All the Dagger implementation is neat and really easy to follow, I like it

samcackett commented on app/src/main/java/com/moonpig/mvisample/di/productdetail/ProductDetailsModule.kt:
```diff
@@ -0,0 +1,27 @@
+            ProductDetailTracker()
+
+    @Provides
+    fun provideProductDetailsViewModelFactory(
```
There's an alternative way of doing this so you don't need to specify the viewModelFactory in every module. See `com.moonpig.di.viewmodel` package in the Moonpig project

JDurstberger replied:
We decided that for MVI demonstration purposes this is not necessary right now. 
Me might want to add more MVI activities in the future and then add the `DaggerViewModelFactory` but in the current state it could be to complex and confusing to someone who just wants to understand MVI

samcackett commented on app/src/main/java/com/moonpig/mvisample/mvibase/BaseViewModel.kt:
```diff
@@ -28,14 +28,12 @@ abstract class BaseViewModel<I : BaseIntent, A : BaseAction, R : BaseResult, VS
-                    .scan(initialViewState()) { previousViewState, result ->
-                        reduce(previousViewState, result)
-                    }
+                    .doOnNext(tracker::trackIntent)
```
Good spot

samcackett commented on app/src/main/java/com/moonpig/mvisample/productdetail/ProductDetailRenderer.kt:
```diff
@@ -0,0 +1,23 @@
+
+    fun render(view: ProductDetailView, viewState: ProductDetailScreenViewState) {
+        renderLoadingIndicator(view, viewState)
+        renderProductDetail(viewState, view)
```
Can you put the parameter args in the same order?

samcackett commented on app/src/main/res/layout/activity_product_detail.xml:
```diff
@@ -0,0 +1,76 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
```
Views go off the screen when rotated

samcackett replied:
Also, none of these views match constraints, can you change that rather than using match_parent or fixed dimensions?

samcackett commented on app/src/main/res/layout/activity_product_detail.xml:
```diff
@@ -0,0 +1,76 @@
+        android:ellipsize="end"
+        android:maxLines="1"
+        android:textAlignment="center"
+
```
Remove new line

samcackett commented on app/src/main/res/layout/activity_product_detail.xml:
```diff
@@ -0,0 +1,76 @@
+
+    <TextView
+        android:id="@+id/name"
+        android:layout_width="match_parent"
```
Can you make this match constraints and constrain it?

samcackett commented on data/src/main/java/com/moonpig/data/ProductDetailStubRepository.kt:
```diff
@@ -1,15 +1,24 @@
+                                    "Advanced cat fighting gift set",
+                                    "Everything your kitty needs to seriously up it's fighting game! Accessories including the pictured mace not included.",
+                                    299,
+                                    "https://d1xkhapf8f3lxw.cloudfront.net/uk/globalassets/gadgets-novelty/update-2/for-him_276x276.jpg?version=4a8967"
```
Can you find a picture of an advanced cat fighting gift set to go here? :trollface:

samcackett reviewed (changes requested):
Testing:
- When rotating the phone, the view goes off the edge and isn't scrollable. We should probably disable rotation or support it

ProductDetailUseCaseTest is broken, imports and classnames missing

Commit:
Fix ProductDetailUseCase test imports

Commit:
Extract dependency versions

Commit:
Fix version names to follow pascalCase naming

Commit:
Reorder ProductDetailRenderer method parameters

Pass view as first and viewState as second parameter

Commit:
Update product detail to show cat armor

Commit:
Add more description to the cat armor

Commit:
Wrap product detail layout in scroll view to support landscape

tyates commented on app/src/androidTest/java/com/moonpig/mvisample/screentests/productdetail/ProductDetailRobot.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    fun isNotLoading() {
+        onView(withId(R.id.progressBar)).check(matches(not(isDisplayed())))
```
There's no scroll so it's not really a problem here but in general I believe we should try to use `matches(withEffectiveVisibility(VISIBLE))` instead of `isDisplayed()` as it avoids the issue where the view is visible but (partially) off the screen.

tyates commented on app/src/androidTest/java/com/moonpig/mvisample/screentests/productdetail/ProductDetailRobot.kt:
```diff
@@ -0,0 +1,31 @@
+
+class ProductDetailRobot {
+    fun isLoading() {
+        onView(withId(R.id.progressBar)).check(matches(isDisplayed()))
```
These two should probably return the Robot as well, it keeps the Robots usage simple and predictable when calling methods in a chain for a more complicated journey test.

tyates replied:
Usually you can't check if these are displayed because they animate and it hogs the render thread so Espresso never resumes until it goes away.

tyates commented on app/src/main/java/com/moonpig/mvisample/productdetail/ProductDetailScreenViewState.kt:
```diff
@@ -0,0 +1,19 @@
+                                  val price: Int = 0,
+                                  val imageUrl: String = ""){
+    companion object {
+        val None = ProductDetailViewState()
```
Not too keen on this as the None state has fields that aren't meaningful. A sealed class with Product and None as children enforces the contract that this can't be displayed.

tyates replied:
You can alternatively make it nullable and use a `?.` call when rendering it, given `None` has no state at all.

tyates commented on app/src/main/res/layout/activity_product_detail.xml:
```diff
@@ -0,0 +1,84 @@
+
+        <ImageView
+            android:id="@+id/image"
+            android:layout_width="200dp"
```
I'd normally say to put this is resources but maybe for the sample we can skip it. Up to you.

JDurstberger replied:
We agreed, that for the sake of the demo this is not important enough.

tyates commented on app/src/main/res/layout/activity_product_detail.xml:
```diff
@@ -0,0 +1,84 @@
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/name"
+            tools:text="This is a long description that is also allowed to wrap into multiple lines"
```
Can use `@tools:sample/lorem/random` or something like that to make some sexy Latin placeholder text.

Commit:
Replace isDisplayed with effective visibility check on productDetailRobot

Commit:
Move progress bar to front of layout

Commit:
Return robot from is(Not)Loading methods

Commit:
Use sample lorem for product detail layout preview

Commit:
Make productDetailViewModel nullable

Todd and I agreed that a simple ?.let is probably the way to go
for that scenario and that we would use a seald class
if the other state would have sensible data to render

tyates reviewed (approved):
There's a warning in build.gradle about a version mismatch for support lib, presumably some transient dependency causing it, can you try to fix it in the error ticket?
